### PR TITLE
Improve SPI library code understanding

### PR DESCRIPTION
**Pull request by fprwi6labs**
Try to improve the clarity of the SPI library code after the feedback #157

### CODE REVIEW

**Comment by fpistm** [comment]
Thanks @fprwi6labs.
Will review it soon

File: libraries/SPI/README.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+## SPI
+
+STM32 SPI library has been modified with the possibility to manage several CS pins without to stop the SPI interface.  
+_We do not describe here the [SPI Arduino API](https://www.arduino.cc/en/Reference/SPI) but the functionalities added._  
+
+We give to the user 3 possiblities about the management of the CS pin:  
+* the CS pin is managed directly by the user code before to transfer the data (like the Arduino SPI library)
```
Comment (fpistm) [inline libraries/SPI/README.md]: numbering:
1.
2.
3.

Remove the 'or'.

File: libraries/SPI/README.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+_We do not describe here the [SPI Arduino API](https://www.arduino.cc/en/Reference/SPI) but the functionalities added._  
+
+We give to the user 3 possiblities about the management of the CS pin:  
+* the CS pin is managed directly by the user code before to transfer the data (like the Arduino SPI library)  
+* or the user gives the CS pin number to the library API and the library manages itself the CS pin (see example below)  
+* or the user uses a hardware CS pin linked to the SPI peripheral  
+
+### New API functions
+
+* **SPIClass::SPIClass(uint8_t mosi, uint8_t miso, uint8_t sclk, uint8_t ssel)**: alternative class constructor
```
Comment (fpistm) [inline libraries/SPI/README.md]: Please, use code formatter 
* **`SPIClass::SPIClass(uint8_t mosi, uint8_t miso, uint8_t sclk, uint8_t ssel)`**: alternative class constructor

File: libraries/SPI/README.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+### New API functions
+
+* **SPIClass::SPIClass(uint8_t mosi, uint8_t miso, uint8_t sclk, uint8_t ssel)**: alternative class constructor  
+_Params_ SPI mosi pin  
+_Params_ SPI miso pin  
+_Params_ SPI sclk pin  
+_Params_ (optional) SPI ssel pin. This pin must be an hardware CS pin. If you configure this pin, the chip select will be managed by the SPI peripheral. Do not use API functions with CS pin in parameter.  
+
+* **void SPIClass::begin(uint8_t _pin)**: initialize the SPI interface and add a CS pin
```
Comment (fpistm) [inline libraries/SPI/README.md]: ditto

File: libraries/SPI/README.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+* **SPIClass::SPIClass(uint8_t mosi, uint8_t miso, uint8_t sclk, uint8_t ssel)**: alternative class constructor  
+_Params_ SPI mosi pin  
+_Params_ SPI miso pin  
+_Params_ SPI sclk pin  
+_Params_ (optional) SPI ssel pin. This pin must be an hardware CS pin. If you configure this pin, the chip select will be managed by the SPI peripheral. Do not use API functions with CS pin in parameter.  
+
+* **void SPIClass::begin(uint8_t _pin)**: initialize the SPI interface and add a CS pin  
+_Params_ spi CS pin to be managed by the SPI library  
+
+* **void beginTransaction(uint8_t pin, SPISettings settings)**: allows to configure the SPI with other parameter. These new parameter are saved this an associated CS pin.
```
Comment (fpistm) [inline libraries/SPI/README.md]: ditto

File: libraries/SPI/README.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+_Params_ (optional) SPI ssel pin. This pin must be an hardware CS pin. If you configure this pin, the chip select will be managed by the SPI peripheral. Do not use API functions with CS pin in parameter.  
+
+* **void SPIClass::begin(uint8_t _pin)**: initialize the SPI interface and add a CS pin  
+_Params_ spi CS pin to be managed by the SPI library  
+
+* **void beginTransaction(uint8_t pin, SPISettings settings)**: allows to configure the SPI with other parameter. These new parameter are saved this an associated CS pin.  
+_Params_ SPI CS pin to be managed by the SPI library  
+_Params_ SPI settings  
+
+* **void endTransaction(uint8_t pin)**: removes a CS pin and the SPI settings associated
```
Comment (fpistm) [inline libraries/SPI/README.md]: ditto

File: libraries/SPI/README.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+_Params_ spi CS pin to be managed by the SPI library  
+
+* **void beginTransaction(uint8_t pin, SPISettings settings)**: allows to configure the SPI with other parameter. These new parameter are saved this an associated CS pin.  
+_Params_ SPI CS pin to be managed by the SPI library  
+_Params_ SPI settings  
+
+* **void endTransaction(uint8_t pin)**: removes a CS pin and the SPI settings associated  
+_Params_ SPI CS pin managed by the SPI library  
+
+**_Note 1_** The following functions must be called after initialization of the SPI instance with begin() or beginTransaction().
```
Comment (fpistm) [inline libraries/SPI/README.md]: Use code formatter for function:
`begin()` or `beginTransaction()`

File: libraries/SPI/README.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+_Params_ SPI settings  
+
+* **void endTransaction(uint8_t pin)**: removes a CS pin and the SPI settings associated  
+_Params_ SPI CS pin managed by the SPI library  
+
+**_Note 1_** The following functions must be called after initialization of the SPI instance with begin() or beginTransaction().  
+If you have several device to manage, you can call beginTransaction() several time with different CS pin in parameter.  
+Then you can call the following functions with different CS pin without call again beginTransaction() (until you call end() or endTransaction()).  
+
+**_Note 2_** If the mode is set to SPI_CONTINUE, the CS pin is kept enabled. Be careful in case you use several CS pin.
```
Comment (fpistm) [inline libraries/SPI/README.md]: Use code formatter for CONSTANT:
`SPI_CONTINUE`

File: libraries/SPI/README.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+* **void endTransaction(uint8_t pin)**: removes a CS pin and the SPI settings associated  
+_Params_ SPI CS pin managed by the SPI library  
+
+**_Note 1_** The following functions must be called after initialization of the SPI instance with begin() or beginTransaction().  
+If you have several device to manage, you can call beginTransaction() several time with different CS pin in parameter.  
+Then you can call the following functions with different CS pin without call again beginTransaction() (until you call end() or endTransaction()).  
+
+**_Note 2_** If the mode is set to SPI_CONTINUE, the CS pin is kept enabled. Be careful in case you use several CS pin.  
+
+* **byte transfer(uint8_t pin, uint8_t _data, SPITransferMode _mode = SPI_LAST)**: write/read one byte
```
Comment (fpistm) [inline libraries/SPI/README.md]: Use code formatter for function:

File: libraries/SPI/README.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+**_Note 2_** If the mode is set to SPI_CONTINUE, the CS pin is kept enabled. Be careful in case you use several CS pin.  
+
+* **byte transfer(uint8_t pin, uint8_t _data, SPITransferMode _mode = SPI_LAST)**: write/read one byte  
+_Params_ SPI CS pin managed by the SPI library  
+_Params_ data to write  
+_Params_ (optional) if SPI_LAST CS pin is reset, SPI_CONTINUE the CS pin is kept enabled.  
+_Return_ byte received  
+
+* **uint16_t transfer16(uint8_t pin, uint16_t _data, SPITransferMode _mode = SPI_LAST)**: write/read half-word
```
Comment (fpistm) [inline libraries/SPI/README.md]: ditto

File: libraries/SPI/README.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+_Params_ (optional) if SPI_LAST CS pin is reset, SPI_CONTINUE the CS pin is kept enabled.  
+_Return_ byte received  
+
+* **uint16_t transfer16(uint8_t pin, uint16_t _data, SPITransferMode _mode = SPI_LAST)**: write/read half-word  
+_Params_ SPI CS pin managed by the SPI library  
+_Params_ 16bits data to write  
+_Params_ (optional) if SPI_LAST CS pin is reset, SPI_CONTINUE the CS pin is kept enabled.  
+_Return_ 16bits data received  
+
+* **void transfer(uint8_t pin, void *_buf, size_t _count, SPITransferMode _mode = SPI_LAST)**: write/read several bytes. Only one buffer used to write and read the data
```
Comment (fpistm) [inline libraries/SPI/README.md]: Use code formatter for function

File: libraries/SPI/README.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+* **byte transfer(uint8_t pin, uint8_t _data, SPITransferMode _mode = SPI_LAST)**: write/read one byte  
+_Params_ SPI CS pin managed by the SPI library  
+_Params_ data to write  
+_Params_ (optional) if SPI_LAST CS pin is reset, SPI_CONTINUE the CS pin is kept enabled.  
+_Return_ byte received  
+
+* **uint16_t transfer16(uint8_t pin, uint16_t _data, SPITransferMode _mode = SPI_LAST)**: write/read half-word  
+_Params_ SPI CS pin managed by the SPI library  
+_Params_ 16bits data to write  
+_Params_ (optional) if SPI_LAST CS pin is reset, SPI_CONTINUE the CS pin is kept enabled.
```
Comment (fpistm) [inline libraries/SPI/README.md]: Use code formatter for CONSTANT

File: libraries/SPI/README.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+_Params_ SPI CS pin managed by the SPI library  
+_Params_ 16bits data to write  
+_Params_ (optional) if SPI_LAST CS pin is reset, SPI_CONTINUE the CS pin is kept enabled.  
+_Return_ 16bits data received  
+
+* **void transfer(uint8_t pin, void *_buf, size_t _count, SPITransferMode _mode = SPI_LAST)**: write/read several bytes. Only one buffer used to write and read the data  
+_Params_ SPI CS pin managed by the SPI library  
+_Params_ pointer to data to write. The data will be replaced by the data read.  
+_Params_ number of data to write/read.
+_Params_ (optional) if SPI_LAST CS pin is reset, SPI_CONTINUE the CS pin is kept enabled.
```
Comment (fpistm) [inline libraries/SPI/README.md]: Use code formatter for CONSTANT

File: libraries/SPI/README.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+_Params_ (optional) if SPI_LAST CS pin is reset, SPI_CONTINUE the CS pin is kept enabled.  
+_Return_ 16bits data received  
+
+* **void transfer(uint8_t pin, void *_buf, size_t _count, SPITransferMode _mode = SPI_LAST)**: write/read several bytes. Only one buffer used to write and read the data  
+_Params_ SPI CS pin managed by the SPI library  
+_Params_ pointer to data to write. The data will be replaced by the data read.  
+_Params_ number of data to write/read.
+_Params_ (optional) if SPI_LAST CS pin is reset, SPI_CONTINUE the CS pin is kept enabled.  
+
+* **void transfer(byte _pin, void *_bufout, void *_bufin, size_t _count, SPITransferMode _mode = SPI_LAST)**: write/read several bytes. One buffer for the output data and one for the input data
```
Comment (fpistm) [inline libraries/SPI/README.md]: Use code formatter for function

File: libraries/SPI/src/SPI.cpp [hunk 14]
```diff
@@ -11,44 +11,52 @@
...
   _spi.pin_sclk = digitalPinToPinName(SCK);
   _spi.pin_ssel = NC;
 }
 
-/* By default hardware SS pin is not used. To use hardware SS pin you should set
-ssel pin. Enable this pin disable software CS. See microcontroller documentation
-for the list of available SS pins. */
-SPIClass::SPIClass(uint8_t mosi, uint8_t miso, uint8_t sclk, uint8_t ssel) : _CSpin(-1)
+/**
+  * @brief  Constructor to create another spi instance attach to another spi
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI
attached

File: libraries/SPI/src/SPI.cpp [hunk 15]
```diff
@@ -11,44 +11,52 @@
...
+  *         peripheraldifferent of the default SPI. All pins must be attached to
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: missing space:
peripheraldifferent

File: libraries/SPI/README.md [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+_Params_ pointer to data to write. The data will be replaced by the data read.  
+_Params_ number of data to write/read.
+_Params_ (optional) if SPI_LAST CS pin is reset, SPI_CONTINUE the CS pin is kept enabled.  
+
+* **void transfer(byte _pin, void *_bufout, void *_bufin, size_t _count, SPITransferMode _mode = SPI_LAST)**: write/read several bytes. One buffer for the output data and one for the input data  
+_Params_ SPI CS pin managed by the SPI library  
+_Params_ pointer to data to write.  
+_Params_ pointer where to store the data read.  
+_Params_ number of data to write/read.  
+_Params_ (optional) if SPI_LAST CS pin is reset, SPI_CONTINUE the CS pin is kept enabled.
```
Comment (fpistm) [inline libraries/SPI/README.md]: Use code formatter for CONSTANT

File: libraries/SPI/src/SPI.cpp [hunk 17]
```diff
@@ -11,44 +11,52 @@
...
+  *         the same spi peripheral. See datasheet of the microcontroller.
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 18]
```diff
@@ -11,44 +11,52 @@
...
+  * @param  mosi: spi mosi pin. Accepted format: decimal number or Arduino format (Dx)
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: ditto
hex or binary is also number

File: libraries/SPI/src/SPI.cpp [hunk 19]
```diff
@@ -11,44 +11,52 @@
...
+  *         or ST format (Pxy).
+  * @param  miso: spi miso pin. Accepted format: decimal number or Arduino format (Dx)
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: ditto

File: libraries/SPI/src/SPI.cpp [hunk 20]
```diff
@@ -11,44 +11,52 @@
...
+  *         or ST format (Pxy).
+  * @param  sclk: spi clock pin. Accepted format: decimal number or Arduino format (Dx)
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: ditto

File: libraries/SPI/src/SPI.cpp [hunk 21]
```diff
@@ -11,44 +11,52 @@
...
+  *         or ST format (Pxy).
+  * @param  ssel: spi ssel pin (optional). Accepted format: decimal number or
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: ditto

File: libraries/SPI/src/SPI.cpp [hunk 22]
```diff
@@ -11,44 +11,52 @@
...
+  *         Arduino format (Dx) or ST format (Pxy). By default is set to NC.
+  *         This pin must correspond to a hardware CS pin which can be managed
+  *         by the spi peripheral itself. See the datasheet of the microcontroller
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 23]
```diff
@@ -11,44 +11,52 @@
...
-{
-  begin(CS_PIN_CONTROLLED_BY_USER);
+  _spi.pin_ssel = digitalPinToPinName(ssel);
 }
 
-//Begin with a chip select defined
+/**
+  * @brief  Initialize the spi instance.
+  * @param  _pin: chip select pin (optional). If this parameter is filled,
+  *         it gives the management of the CS pin to the spi class. In this case
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 24]
```diff
@@ -79,12 +87,16 @@ void SPIClass::begin(uint8_t _pin)
...
-{
-  UNUSED(interruptNumber);
-  //Not implemented
-}
-
+/**
+  * @brief  This function should be used to configure the SPI instance in case you
+  *         don't use the default parameters set by the begin() function.
+  * @param  _pin: CS pin (optional). This pin will be attached with the settings.
+  * @param  settings: spi settings(clock speed, bit order, data mode).
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 25]
```diff
@@ -79,12 +87,16 @@ void SPIClass::begin(uint8_t _pin)
...
+  * @Note   For each spi instance you are able to manage until NB_SPI_SETTINGS
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 26]
```diff
@@ -218,29 +264,35 @@ void SPIClass::setClockDivider(uint8_t _pin, uint8_t _divider)
...
 
-
-/* Transfer a message on the selected SPI. The _pin is the CS.
- * The transfer function can reconfigure the SPI instance if the CS pin is
- * different from the previous one.
- * If the _mode is set to SPI_CONTINUE, keep the spi instance alive. That means
- * the CS pin is not reset. Be careful in case you use several CS pin.
- */
+/**
+  * @brief  Transfer one byte on the spi bus.
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 27]
```diff
@@ -218,29 +264,35 @@ void SPIClass::setClockDivider(uint8_t _pin, uint8_t _divider)
...
+  *         begin() or beginTransaction() must be called at least once before.
+  * @param  _pin: CS pin to select a device (optional). If the previous transfer
+  *         used another CS pin then the spi instance will be reconfigured.
+  * @param  data: byte to send.
+  * @param  _mode: (optional) can be SPI_CONTINUE in case of multiple successive
+  *         send or SPI_LAST to indicate the end of send.
+  *         If the _mode is set to SPI_CONTINUE, keep the spi instance alive.
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 28]
```diff
@@ -254,6 +306,19 @@ byte SPIClass::transfer(uint8_t _pin, uint8_t data, SPITransferMode _mode)
   return rx_buffer;
 }
 
+/**
+  * @brief  Transfer two bytes on the spi bus in 16 bits format.
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 29]
```diff
@@ -254,6 +306,19 @@ byte SPIClass::transfer(uint8_t _pin, uint8_t data, SPITransferMode _mode)
...
+  *         begin() or beginTransaction() must be called at least once before.
+  * @param  _pin: CS pin to select a device (optional). If the previous transfer
+  *         used another CS pin then the spi instance will be reconfigured.
+  * @param  data: bytes to send.
+  * @param  _mode: (optional) can be SPI_CONTINUE in case of multiple successive
+  *         send or SPI_LAST to indicate the end of send.
+  *         If the _mode is set to SPI_CONTINUE, keep the spi instance alive.
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 30]
```diff
@@ -295,20 +360,34 @@ uint16_t SPIClass::transfer16(uint8_t _pin, uint16_t data, SPITransferMode _mode
...
+  * @brief  Transfer several bytes. Only one buffer used to send and receive data.
+  *         begin() or beginTransaction() must be called at least once before.
+  * @param  _pin: CS pin to select a device (optional). If the previous transfer
+  *         used another CS pin then the spi instance will be reconfigured.
+  * @param  _buf: pointer to the bytes to send. The bytes received are copy in
+  *         this buffer.
+  * @param  _count: number of bytes to send/receive.
+  * @param  _mode: (optional) can be SPI_CONTINUE in case of multiple successive
+  *         send or SPI_LAST to indicate the end of send.
+  *         If the _mode is set to SPI_CONTINUE, keep the spi instance alive.
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 31]
```diff
@@ -320,20 +399,35 @@ void SPIClass::transfer(uint8_t _pin, void *_buf, size_t _count, SPITransferMode
...
+  *         another one will contains the data received. begin() or
+  *         beginTransaction() must be called at least once before.
+  * @param  _pin: CS pin to select a device (optional). If the previous transfer
+  *         used another CS pin then the spi instance will be reconfigured.
+  * @param  _bufout: pointer to the bytes to send.
+  * @param  _bufin: pointer to the bytes received.
+  * @param  _count: number of bytes to send/receive.
+  * @param  _mode: (optional) can be SPI_CONTINUE in case of multiple successive
+  *         send or SPI_LAST to indicate the end of send.
+  *         If the _mode is set to SPI_CONTINUE, keep the spi instance alive.
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 32]
```diff
@@ -320,20 +399,35 @@ void SPIClass::transfer(uint8_t _pin, void *_buf, size_t _count, SPITransferMode
     digitalWrite(_pin, HIGH);
 }
 
+/**
+  * @brief  Transfer several bytes. One buffer contains the data to send and
+  *         another one will contains the data received. begin() or
+  *         beginTransaction() must be called at least once before.
+  * @param  _pin: CS pin to select a device (optional). If the previous transfer
+  *         used another CS pin then the spi instance will be reconfigured.
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.cpp [hunk 33]
```diff
@@ -295,20 +360,34 @@ uint16_t SPIClass::transfer16(uint8_t _pin, uint16_t data, SPITransferMode _mode
   return rx_buffer;
 }
 
+/**
+  * @brief  Transfer several bytes. Only one buffer used to send and receive data.
+  *         begin() or beginTransaction() must be called at least once before.
+  * @param  _pin: CS pin to select a device (optional). If the previous transfer
+  *         used another CS pin then the spi instance will be reconfigured.
```
Comment (fpistm) [inline libraries/SPI/src/SPI.cpp]: spi -> SPI

File: libraries/SPI/src/SPI.h [hunk 34]
```diff
@@ -137,63 +162,35 @@ class SPIClass {
...
-    void setBitOrder(BitOrder _order) {
-      setBitOrder(CS_PIN_CONTROLLED_BY_USER, _order);
-    }
-    void setDataMode(uint8_t _mode) {
-      setDataMode(CS_PIN_CONTROLLED_BY_USER, _mode);
-    }
-    void setClockDivider(uint8_t _div) {
-      setClockDivider(CS_PIN_CONTROLLED_BY_USER, _div);
-    }
+    // Not implemented functions. Keep for compatibility.
```
Comment (fpistm) [inline libraries/SPI/src/SPI.h]: backward compatibility

**Comment by fprwi6labs** [comment]
Done

File: libraries/SPI/src/SPI.h [hunk 35]
```diff
@@ -137,63 +162,35 @@ class SPIClass {
...
-    void setBitOrder(BitOrder _order) {
-      setBitOrder(CS_PIN_CONTROLLED_BY_USER, _order);
-    }
-    void setDataMode(uint8_t _mode) {
-      setDataMode(CS_PIN_CONTROLLED_BY_USER, _mode);
-    }
-    void setClockDivider(uint8_t _div) {
-      setClockDivider(CS_PIN_CONTROLLED_BY_USER, _div);
-    }
+    // Not implemented functions. Keep for backward compatibility.
```
Comment (fpistm) [inline libraries/SPI/src/SPI.h]: Kept

Comment (fprwi6labs) [reply libraries/SPI/src/SPI.h]: done

**Pull request merged**